sophiread/SophireadCLI/src/sophiread_core.cpp

```diff
@@ -0,0 +1,382 @@
+ *
+ * @param[in] batches The vector of TPX3 batches containing the neutron events.
+ * @param[in] super_resolution The super resolution factor used to calculate the dimensions of each 2D histogram.
+ * @param[in] tof_bin_edges The vector of TOF bin edges used to determine the TOF bin for each neutron event.
+ * @param[in] mode The mode of operation, either "hit" or "neutron", which determines the type of events to process.
+ * @return std::vector<std::vector<std::vector<unsigned int>>> The vector of TOF images, where each TOF image is a 2D histogram representing the distribution of neutron events in space for a specific TOF bin.
+ */
+std::vector<std::vector<std::vector<unsigned int>>> timedCreateTOFImages(
+    const std::vector<TPX3>& batches, 
+    double super_resolution, 
+    const std::vector<double>& tof_bin_edges,
+    const std::string& mode) {
+    
+    auto start = std::chrono::high_resolution_clock::now();
+
+    // Initialize the TOF images container
+    std::vector<std::vector<std::vector<unsigned int>>> tof_images(tof_bin_edges.size() - 1);
+
+    // Sanity checks
+    if (tof_bin_edges.size() < 2) {
+        spdlog::error("Invalid TOF bin edges: at least 2 edges are required");
+        return {};
+    }
+    if (batches.empty()) {
+        spdlog::error("No batches to process");
+        return tof_images;
+    }
+    
+    // Calculate the dimensions of each 2D histogram based on super_resolution
+    // one chip: 0-255 pixel pos
+    // gap: 5
+    // total: 0-255 + 5 + 0-255 -> 517 (TPX3@VENUS only)
+    int dim_x = static_cast<int>(517 * super_resolution);
+    int dim_y = static_cast<int>(517 * super_resolution);
+
+    spdlog::debug("Creating TOF images with dimensions: {} x {}", dim_x, dim_y);
+    spdlog::debug("tof_bin_edges size: {}", tof_bin_edges.size());
+    if (!tof_bin_edges.empty()) {
+        spdlog::debug("First bin edge: {}, Last bin edge: {}", tof_bin_edges.front(), tof_bin_edges.back());
+    }
+    
+    // Initialize each TOF bin's 2D histogram
+    for (auto& tof_image : tof_images) {
+        tof_image.resize(dim_y, std::vector<unsigned int>(dim_x, 0));
+    }
+
+    // Process neutrons from all batches
+    size_t total_entries = 0;
+    size_t binned_entries = 0;
+
+    for (size_t batch_index = 0; batch_index < batches.size(); ++batch_index) {
+        const auto& batch = batches[batch_index];
+        spdlog::debug("Processing batch {}", batch_index);
+
+        std::vector<const IPositionTOF*> entries;
+        if (mode == "hit") {
+            entries.reserve(batch.hits.size());
+            for (const auto& hit : batch.hits) {
+                entries.push_back(static_cast<const IPositionTOF*>(&hit));
```

Comment:
Aren't `hit` and `neutron` both subclasses of `IPositionTOF`? I think the batches should be vectors of pointers too then you can abstract some more of this.

Reply:
I tried and failed. Apparently because hits and neutron have very different attributes, their layout is different, so when I tried to just cast the pointers, I get segmentation fault.
I couldn't figure out how to resolve the segfault issue, and decided to fall back to the copy route, which is slow but always work